.ci/lib.sh

```diff
@@ -206,6 +206,17 @@ delete_stale_docker_resource()
 
 delete_stale_kata_resource()
 {
+	# umount before dir
+	for stale_kata_mount_point in "${stale_kata_mount_point_union[@]}"; do
+		local mount_point_union=$(mount | grep "${stale_kata_mount_point}" | awk '{print $3}')
+		if [ -n "${mount_point_union}" ]; then
+			while IFS='$\n' read mount_point; do
+				[ -n "$(grep "${mount_point}" "/proc/mounts")" ] && sudo umount -R "${mount_point}"
```

Comment:
Whilst this lgtm - is this also indicative of a bug in Kata - presumably we should never be leaving these stale mounts around?
/cc @amshinde @devimc

Reply:
if the container exits normally, it will not leave this resource behind. But if it exited abnormally, like the hypervisor(qemu) doesn't start up, it will cause this stale mounts.

Reply:
@Pennyzct - thx - so, I would still call that a bug :-) Under no circumstances should we normally leave stuff laying around. Particularly if it is something like qemu not starting up, which the runtime should notice, and clean up. Could you open an Issue with the details you have please, and then we can try to find somebody to have a look (/cc some folks who may have looked in the cleanup-on-fail area before @WeiZhang555 @bergwolf )

Reply:
Hi~ @grahamwhaley I've been tracked the point where the network namespace gets cleaned up. it's this func [cleanupNetNS](https://github.com/kata-containers/runtime/blob/master/pkg/katautils/network.go). it will get cleaned up when sandbox fails to create, so including qemu not starting up. ahhhhhhh, it's killing me. 😭🤮

Reply:
@Pennyzct cleanup is **hard** - harder than creation :-)  Do you have a reliable test case that makes this happen? Maybe the exit path is not going through `clearnupNetNS` - maybe the golang program is being killed or something. Going to be easiest to track if we can at least reproduce. I know we have had trouble reproducing such 'leaks' in the past... tricky....

Reply:
@grahamwhaley. After looking through log from` journalctl -t kata-runtime` and `/var/log/syslog`, i found nothing related. that's the pain point. o(TヘTo). I couldn't find the exact test case to produce this leak.
I'm thinking, maybe we don't turn on the `debug` in ci running, so maybe missing some info. 
I'm gonna find a real clean local machine to mimic one ARM CI run, to check out if it's reproducing.
BTW, running CI in bare-metal is really pain in the ass. Especially, recently I'm thinking to include
k8s test cases.
Do you guys think that running kata CI in docker container is a feasible solution? @grahamwhaley @chavafg

Reply:
Hi @Pennyzct ... I'm not sure we can run the CI in a docker container - that would need the ability to run Kata in Docker - which I suspect you could only do with a priv container - which might defeat some of the point... if you do follow that route do keep us in the loop though.
However, can you do nested VMs on your hardware? If so, then I did previously have the metrics baremetal running the CI inside VMs (https://github.com/kata-containers/ci/tree/master/VMs/metrics). Hmm, I suspect we have had this conversation before...

Reply:
oh, one other option, if you are using Packet servers, is to try and fix the 'dynamically deploy packet from Jenkins'. I tried this in the past using the [jclouds](https://wiki.jenkins.io/display/JENKINS/JClouds+Plugin) plugin - as packet has a jclouds integration, but packet did not appear in the plugin dropdown of providers - packet were aware of that at the time I think - I don't know if that has been fixed/progressed though...

Alternatively, I see there is a REST API for packet as well - I suspect that could also be used to deploy Jenkins slaves.
[REST API](https://support.packet.com/kb/articles/deploy-via-the-api)

I'd love to dynamically deploy the metrics machines in packet - I have been pursuing that via our Zuul integration, as there is Zuul/packet integration, and that is also the only way I can store metrics results into the OSF/infra/zuul elasticsearch DB as well....

Reply:
Hi~ @grahamwhaley Sadly, nested virtualization is really a long way on AArch64. 😭. So I'm considering container.;) 
As for privilege problem in kata, I only had a ambiguous concept that we couldn't support privileged kata container, as for why, i'm not clear,  would you like to point me some related issues or PRs? 😋

Reply:
Hi @Pennyzct - I think you are after https://github.com/kata-containers/documentation/blob/master/Limitations.md#docker-run---privileged

Reply:
Hi @jodh-intel thanks! I have skipped the whole related issue. So it couldn't provide all runc supports, just a subset of host devices, including block and vfio devices. So not 100 percentage `--priviledged`? 
@grahamwhaley As far as I know, I'm gonna still give a try on running kata CI on docker container. (✿◕‿◕✿)

Reply:
Hi @Pennyzct @jodh-intel - I think we might be mixing up Docker-in-Kata and Kata-in-Docker... to run the Kata CI system *inside* a container, you would need to use Kata-in-Docker. The above links are talking about running priv containers under Kata, and more like Docker-in-Kata....

So, I think you are wanting to try and run Kata **inside** a container @Pennyzct - and as Kata needs to perform some sudo level system configuration (like disk mounts and network connections), it will need to run in a docker priv container I believe. I suggest probably try it - run a docker priv container, try and install kata into it, and run it. Oh, to do that I guess you will need to be running [DockerInDocker](https://hub.docker.com/_/docker) - there is a strong recommendation there to read the relevant blog posts first...

I did consider some time back if it might be easier to run docker/kata under lxc/ldx rather than docker - but I've never tried any of them...

Reply:
Hi @grahamwhaley I'm not gonna using **_the whole package docker in docker_**. It has been proved to have multiple ugly scenarios. 
I'm trying to use this magical thing:
```
docker run -v /var/run/docker.sock:/var/run/docker.sock --privileged \
           -ti docker
```
this blog has detailed info about running ci with DockerInDocker, [Using Docker-in-Docker for your CI or testing environment? Think twice](https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/)
I have to be extramely careful about events involving removingcontainers, since it mat remove the **_docker_** container. ;)
For now, all is in my imagination. I'll find time to give it a try asap. ;)

Reply:
Any update on this @Pennyzct? :)